### PR TITLE
Add Slop::Result#fetch

### PR DESCRIPTION
**Pull request by giovannibenussi**
Sometimes I want to raise an error when a given key is not present (for avoid typos for example) so this change implements Slop::Result#fetch allows the following:

```ruby
opts = Slop.parse do |o|
  o.string '-name'
end

ARGV #=> --name giovanni

opts[:name]          # giovanni
opts.fetch(:name) # giovanni

opts.fetch(:namr)      # raises KeyError: key not found: 'namr'
opts.fetch(:last_name) # raises KeyError: key not found: 'last_name'
```

`Slop::Result#fetch` returns the value associated to a given key as `Slop::Result#[]`  and `Slop::Result#get` but raises `KeyError` if it does not exist.

### CODE REVIEW

File: lib/slop/result.rb [hunk 1]
```diff
@@ -20,6 +20,11 @@ def [](flag)
     end
     alias get []
 
+    # Returns an options value, raises KeyError if the option does not exist.
```
Comment (olleolleolle) [inline lib/slop/result.rb]: `options` => `option's`

**Comment by olleolleolle** [comment]
[Hash#fetch](https://ruby-doc.org/core-2.5.1/Hash.html#method-i-fetch) implements a few more behaviors, are you interested in those, too?

File: lib/slop/result.rb (hunk 1 above)
Comment (giovannibenussi) [reply lib/slop/result.rb]: Thanks, fixed here and in `Slop::Result#[]` description!

**Comment by giovannibenussi** [comment]
I take a look at it but I think that is a good option to not allow a default value when a given key does not exists, primarily because you know beforehand all the keys that you'll have and you can also specify default values for those keys. What do you think @olleolleolle?

**Comment by olleolleolle** [comment]
@giovannibenussi I usually raise a more specific error message or specific error class in `Hash#fetch`'s block.

**Comment by olleolleolle** [comment]
(But, this was a question, not a judgement.)

**Comment by giovannibenussi** [comment]
No worries :-) !

I'll take `ActiveSupport::OrderedOptions` as a reference, where they implement `options.attribute` and `options.attribute!`. I was taking a look at the pull request when the bang version was proposed and I think that this comment from @asanghi summarises it pretty good:

> secrets.yml is more likely than not to contain passwords/keys to services which might be essential to your application's execution. Without those keys the application is unlikely to work as expected (or crash in unexpected ways). The bang version of the method basically allows us to crash in a more expected way, letting us know that a blank has been encountered when it's not expected. In absence of this, we're more likely to write custom code checking for blanks and handling it with custom error methods (and then possibly erroring out anyway).

For the case of `Hash#fetch`, I would use a default value when I don't know beforehand all the keys that will be processed and I know how to gracefully handle those cases in a generalized way. Slop is different because you know beforehand all the keys and if one is not set, then you can specify a default value.

Let me know you thoughts :-)

**Comment by giovannibenussi** [comment]
I notice that if you declare an option as not required without a default value and don't set it, then `Slop::Result#fetch` will raise an error:

```ruby
opts = Slop.parse do |o|
  o.string '-name'
end

ARGV #=> ''

opts.fetch(:name)      # raises KeyError: key not found: 'name'
```

I think that in this case then `nil` should be returned.
I can work on this later!

**Comment by giovannibenussi** [comment]
I made some changes so now `Slop::Result#fetch` handles the case where a key is not provided and it does not have a default value. I also add a test for this case and the case when a value is not provided but the option has a default value.

File: lib/slop/result.rb [hunk 2]
```diff
@@ -14,12 +14,22 @@ def initialize(parser)
...
     def [](flag)
       (o = option(flag)) && o.value
     end
     alias get []
 
+    # Returns an option's value, raises KeyError if the option does not exist.
+    def fetch(flag)
+      o = option(flag)
+      if o.nil?
+        raise(KeyError.new(%Q(key not found: '#{clean_key(flag)}')))
```
Comment (olleolleolle) [inline lib/slop/result.rb]: can also be expressed as

`raise KeyError, %Q(key not found: '#{clean_key(flag)}')`

Comment (giovannibenussi) [reply lib/slop/result.rb]: What about:

`raise KeyError, "key not found: '#{clean_key(flag)}'"`

I think that I was using `%Q` because at some point I wanted to show an error message like `key not found: "key"` but it is no longer necessary.

**Comment by giovannibenussi** [comment]
How do you feel about this changes?
They are ok for you? :-)

Regards!

File: lib/slop/result.rb [hunk 3]
```diff
@@ -14,12 +14,22 @@ def initialize(parser)
...
-    # Returns an options value, nil if the option does not exist.
+    # Returns an option's value, nil if the option does not exist.
     def [](flag)
       (o = option(flag)) && o.value
     end
     alias get []
 
+    # Returns an option's value, raises KeyError if the option does not exist.
+    def fetch(flag)
+      o = option(flag)
```
Comment (leejarvis) [inline lib/slop/result.rb]: This could be simpler because we already support the functionality in `Result#[]`:

```ruby
def fetch(flag)
  get(flag) or raise KeyError, "key not found: '#{clean_key(flag)}'"
end
```

Also I don't think the message is clear enough. I would opt for something like "Could not find option with flag '#{flag}'". I don't think the flag needs cleaning, either, but I'm not too bothered either way. Also, any reason we shouldn't just re-use `Slop::UnknownOption` since it basically means the same thing?

File: lib/slop/result.rb [hunk 4]
```diff
@@ -33,16 +43,17 @@ def []=(flag, value)
...
-        key = key.tr '-', '_' if parser.config[:underscore_flags]
-        key.to_sym
-      end
       options.find do |o|
-        o.flags.any? { |f| cleaned.(f) == cleaned.(flag) }
+        o.flags.any? { |f| clean_key(f) == clean_key(flag) }
       end
     end
 
+    def clean_key(key)
```
Comment (leejarvis) [inline lib/slop/result.rb]: If you decide to keep this method, please move it to the bottom of the class and mark it as `private`. It should only be used internally

File: test/result_test.rb [hunk 5]
```diff
@@ -62,6 +62,51 @@ def finish(result)
...
   end
 
+  describe "#fetch" do
+    it "returns an options value" do
+      assert_equal "lee", @result.fetch("name")
+      assert_equal "lee", @result.fetch(:name)
+      assert_equal "lee", @result.fetch("--name")
+      assert_equal "bar", @result.fetch("long_option")
+      assert_equal "bar", @result.fetch(:long_option)
+      assert_equal "bar", @result.fetch("--long-option")
```
Comment (leejarvis) [inline test/result_test.rb]: I think this is a bit over-tested because it's not just testing `Result#fetch`, it's also testing `Result#get` and others. I think the test examples are good though, but all of them only need one assertion without the duplication

File: lib/slop/result.rb (hunk 3 above)
Comment (giovannibenussi) [reply lib/slop/result.rb]: `get(flag) or raise KeyError, "key not found: '#{clean_key(flag)}'"` will throw an exception if the value of the key `flag` is nil this happens when you don't provide an option and that option does not have a default value as I test in _test/result_test.rb_:

```ruby
it "returns nil when an option is not provided and it does not have a default value" do
  @options.string("--hello")
  @result.parser.parse %w()

  assert_equal nil, @result.fetch('hello')
end
```

I tried to remove `clean_key(flag)` but if you try to make something like `@result.fetch("--unexisting")` then you'll receive the following error: `key not found: '--unexisting'`.

Comment (leejarvis) [reply lib/slop/result.rb]: > will throw an exception if the value of the key flag is nil this happens when you don't provide an option and that option does not have a default value as I test in test/result_test.rb

Right, good catch! 

> I tried to remove clean_key(flag) but if you try to make something like @result.fetch("--unexisting") then you'll receive the following error: key not found: '--unexisting'.

Yeah, I think that's okay. As I mentioned though, I don't feel too strongly about it so feel free to keep it. I'd be up for re-using `Slop::UnknownOption`, though

Comment (giovannibenussi) [reply lib/slop/result.rb]: > I'd be up for re-using Slop::UnknownOption, though

I think about it and you're right, so I make the changes for this!

I update the documentation for `Slop::UnknownOption` also so maybe you want to take a look at it:

> Raised when an unknown option is parsed or when trying to fetch an
unexisting option via `Slop::Result#fetch`.
Suppress with the `suppress_errors` config option.

File: lib/slop/result.rb (hunk 4 above)
Comment (giovannibenussi) [reply lib/slop/result.rb]: Done!

File: test/result_test.rb (hunk 5 above)
Comment (giovannibenussi) [reply test/result_test.rb]: You think it's ok to just leave the following two tests:

```ruby
assert_equal "lee", @result.fetch("name")
```

and

```ruby
e = assert_raises(KeyError) { @result.fetch("unexisting") }
assert_equal "key not found: 'unexisting'", e.message
```

I'm not pretty sure if remove `"returns the default value of an option when a value is not provided"` and `"returns nil when an option is not provided and it does not have a default value"`, because they test edge cases, especially the last one.

Comment (leejarvis) [reply test/result_test.rb]: > I'm not pretty sure if remove "returns the default value of an option when a value is not provided" and "returns nil when an option is not provided and it does not have a default value", because they test edge cases, especially the last one.

I think they should be kept too, yes, especially since they catch the `nil` issue I mentioned in another comment :)

Comment (giovannibenussi) [reply test/result_test.rb]: Cool, I just push those changes :-)

File: test/result_test.rb [hunk 6]
```diff
@@ -64,12 +64,12 @@ def finish(result)
...
     it "returns an options value" do
-      assert_equal "lee", @result.fetch("name")
-      assert_equal "lee", @result.fetch(:name)
       assert_equal "lee", @result.fetch("--name")
-      assert_equal "bar", @result.fetch("long_option")
-      assert_equal "bar", @result.fetch(:long_option)
-      assert_equal "bar", @result.fetch("--long-option")
+    end
+
+    it "raises KeyError when an option does not exists" do
```
Comment (leejarvis) [inline test/result_test.rb]: The description needs updating now :)

**Comment by olleolleolle** [comment]
This is a fine addition to Slop!

Collaboration is happening!

Thanks for creating this, together!

File: test/result_test.rb (hunk 6 above)
Comment (giovannibenussi) [reply test/result_test.rb]: Good catch, updated!

File: lib/slop/result.rb [hunk 7]
```diff
@@ -14,12 +14,23 @@ def initialize(parser)
...
       (o = option(flag)) && o.value
     end
     alias get []
 
+    # Returns an option's value, raises KeyError if the option does not exist.
+    def fetch(flag)
+      o = option(flag)
+      if o.nil?
+        cleaned_key = clean_key(flag)
+        raise UnknownOption.new("key not found: '#{cleaned_key}'", "#{cleaned_key}")
```
Comment (leejarvis) [inline lib/slop/result.rb]: One last nitpick I promise :) 

I think the error message should have more detail. "key not found" might be confusing in the context of command line options and flags. After all, what is a key? Perhaps we could just re-use `option` like we did for the other `UnknownOption` class ("option not found") to clear this up? That might be enough to avoid confusion. What do you think?

**Comment by leejarvis** [comment]
@giovannibenussi Thanks for all your work and patience! This looks really great. Please could you squash all of your commits into one? (or into as many as you'd like if you think they're worth retaining for context). Then we'll get it merged. Thanks again! ✨ 💜

File: lib/slop/result.rb (hunk 7 above)
Comment (giovannibenussi) [reply lib/slop/result.rb]: Don't worry, I also really want to have this feature as good as possible so don't worry!
If you have any other suggestion I'm open to them :-)

Thanks for your feedback!

**Comment by giovannibenussi** [comment]
Sure @leejarvis, I'm gonna make it later because I don't have that much experience squashing commits  but I'm gonna learn how to do it!

**Comment by leejarvis** [comment]
@giovannibenussi Great! Give me a shout on Twitter at lee_jarvis or if you're on IRC then irc.freenode.org in the `#ruby-lang` channel as ljarvis and I'll help you go through it

**Comment by olleolleolle** [comment]
@giovannibenussi I'm partial to [git-extras](https://github.com/tj/git-extras)' `git squash` command. [Documentation link](https://github.com/tj/git-extras/blob/master/man/git-squash.md)

---

@leejarvis If you interact with the tiny white triangle on the huge green Merge button, _you_ can pick _Squash and Merge_, but then you won't get any "Merge Commit", which is a difference, of course. Do you prefer to have Merge Commits?

image

**Comment by giovannibenussi** [comment]
Thanks for your help guys, I finally can squash the commits with the help of a [great tutorial](https://brunoscopelliti.com/squash-commits-with-git-rebase) that I find out!

Thanks a lot for your help and guidance through this new feature, it was nice to work with both of you!

If you have another last-minute change just tell me!
And if you need help with anything I'm open to help you guys :-)

Regards!

**Comment by leejarvis** [comment]
@olleolleolle Thanks, I didn't know of that button addition.

@giovannibenussi Great work! Thanks again for the contribution, I'll get this merged before the end of the week (@olleolleolle feel free to merge and update the changelog if you'd like)

**Comment by giovannibenussi** [comment]
No problem @leejarvis !

Thanks to both of you again! :-)

**Comment by olleolleolle** [comment]
There, CHANGELOG amended.
